src/algebra/category/Group/biproducts.lean

```diff
@@ -0,0 +1,186 @@
+      ext; [rw ← w walking_pair.left, rw ← w walking_pair.right]; refl,
+    end, } }
+
+instance has_colimit_pair (G H : AddCommGroup.{u}) : has_colimit.{u} (pair G H) :=
+{ cocone :=
+  { X := AddCommGroup.of (G × H),
+    ι := { app := λ j, walking_pair.cases_on j (add_monoid_hom.inl G H) (add_monoid_hom.inr G H) }},
+  is_colimit :=
+  { desc := λ s, add_monoid_hom.coprod (s.ι.app walking_pair.left) (s.ι.app walking_pair.right),
+    fac' :=
+    begin
+      rintros s (⟨⟩|⟨⟩); { ext x, dsimp, simp, erw [add_monoid_hom.map_zero], refl, }
+    end,
+    uniq' := λ s m w,
+    begin
+      ext,
+      rw [←w, ←w],
+      simp only [binary_bicone.to_cocone, add_monoid_hom.coprod_apply, coe_comp], dsimp,
+      simp only [add_monoid_hom.inl_apply, add_monoid_hom.inr_apply],
+      erw [←add_monoid_hom.map_add],
+      cases x, dsimp,
+      rw prod.mk_add_mk,
+      erw [zero_add, add_zero],
+    end , }, }
+
+instance (G H : AddCommGroup.{u}) : has_binary_biproduct.{u} G H :=
+{ bicone :=
+  { X := AddCommGroup.of (G × H),
+    π₁ := add_monoid_hom.fst G H,
+    π₂ := add_monoid_hom.snd G H,
+    ι₁ := add_monoid_hom.inl G H,
+    ι₂ := add_monoid_hom.inr G H, },
+  is_limit := limit.is_limit (pair G H),
+  is_colimit := colimit.is_colimit (pair G H), }
+
+-- We verify that the underlying type of the biproduct we've just defined is definitionally
+-- the cartesian product of the underlying types:
+example (G H : AddCommGroup.{u}) : ((G ⊞ H : AddCommGroup.{u}) : Type u) = (G × H) := rfl
+
+variables {J : Type u} (F : (discrete J) ⥤ AddCommGroup.{u})
+
+namespace has_limit
+
+/--
+The map from an arbitrary cone over a indexed family of abelian groups
+to the cartesian product of those groups.
+-/
+def lift (s : cone F) :
+  s.X ⟶ AddCommGroup.of (Π j, F.obj j) :=
+{ to_fun := λ x j, s.π.app j x,
+  map_zero' := begin ext, dsimp, erw [add_monoid_hom.map_zero], refl, end, -- FIXME why can't simp do this, using pi.zero_apply?
+  map_add' := λ x y, begin ext, dsimp, erw [add_monoid_hom.map_add], refl, end, }
+
+@[simp] lemma lift_apply (s : cone F) (x : s.X) (j : J) : (lift F s) x j = s.π.app j x := rfl
+
+instance has_limit_discrete : has_limit F :=
+{ cone :=
+  { X := AddCommGroup.of (Π j, F.obj j),
+    π := nat_trans.of_homs (λ j, add_monoid_hom.apply (λ j, F.obj j) j), },
+  is_limit :=
+  { lift := lift F,
+    fac' := λ s j, begin dsimp, ext, dsimp, simp, refl, end,
+    uniq' := λ s m w,
+    begin
+      ext x j,
+      dsimp only [has_limit.lift],
+      simp only [add_monoid_hom.coe_mk],
+      convert congr_arg (λ f : s.X ⟶ F.obj j, (f : s.X → F.obj j) x) (w j),
+    end, }, }
+
+end has_limit
+
+namespace has_colimit
+variables [fintype J]
+
+/--
+The map from the cartesian product of a finite family of abelian groups
+to any cocone over that family.
+-/
+def desc (s : cocone F) :
+  AddCommGroup.of (Π j, F.obj j) ⟶ s.X :=
+{ to_fun := λ f, finset.univ.sum (λ j : J, s.ι.app j (f j)),
+  map_zero' :=
+  begin
+    transitivity,
```

Comment:
This proof could use some curly braces, I think.

Reply:
This proof is just too disgusting; we can't merge as is. I'm crossing my fingers that 3.8 magically helps here, and investigating that now.

Reply:
Well, I managed to make a lot of things better. I'm not 100% sure that it was 3.8 vs being less dumb than last time I looked at this, but let's give 3.8 the credit. :)